posts/2018-01-26-announcing-pidgin-v2.0.md

```diff
@@ -0,0 +1,304 @@
+---
+title: Announcing Pidgin v2.0
+subtitle: Errors, `Span`, and performance
+---
+
+I'm pleased to announce that version 2.0 of my functional parsing library, [Pidgin](https://github.com/benjamin-hodgson/Pidgin), is now available [on Nuget](https://www.nuget.org/packages/Pidgin). In this release I've focused on error messages, performance, and `Span` support.
+
+
+Parser Combinators
+------------------
+
+I haven't written about Pidgin before, so allow me to briefly introduce it. Pidgin is a [_parser combinator_](https://en.wikipedia.org/wiki/Parser_combinator) library, meaning that it consists of three main concepts:
+
+* A type `Parser` which models a _parsing process_
+* A collection of _primitive_ `Parser` objects which perform some simple individual parsing task
+* A collection of _combinator_ functions which can build complex `Parser`s out of simpler ones
+
+Taken together, we have an object model allowing you to compose parsers using an embedded domain-specific langauge. As a brief taste, here is a simple example parser which parses an identifier in a typical programming language.
+
+```csharp
+Parser<char, string> Identifier =
+    Letter.Then(
+        LetterOrDigit.ManyString(),
+        (first, rest) => first + rest
+    );
+
+Assert.Equal("abc1", Identifier.ParseOrThrow("abc1"));
+```
+
+* `Identifier` is a `Parser<char, string>`, meaning it's a parsing process which consumes a sequence of `char`s and produces a `string`.
+* `Letter` and `LetterOrDigit` are primitive parsers, each of which read a single character from the input string (ensuring that the character is alphabetical or alphanumeric, respectively).
+* `ManyString` is a combinator method which runs a parser repeatedly and packs the result into a `string` --- so `LetterOrDigit.ManyString()` is a parser which reads a sequence of alphanumeric characters.
```

Comment:
ok so it sounds like i have a bit more explaining to do. I'll have a go here and if it makes sense I'll try to integrate it into the paragraph:

`LetterOrDigit` is a parser which _consumes_ a single character from the input stream, as long as that character is a letter or a digit. That means the input stream gets advanced to the next character if the current character is a letter or digit. If it's not, then the parser fails. The parser's return value is the character that it consumed.

`ManyString` is a combinator method which runs a parser in a loop until it fails. The `ManyString` parser's return value is all of the results from the loop, packed into a string.

So `LetterOrDigit.ManyString()` is a parser which consumes a sequence of letter-or-digit characters from the input stream, and returns that sequence of characters as a string.

Reply:
I'm thinking now that an introductory "how parsers work" article would be a good follow up post, rather than making this one even longer?

Reply:
That explanation above is great - I get it much better now. An intro how parsers work article would be great too. Totally depends on who you imagine your audience to be. If they are people who already know how parsers work, which I'm assuming they probably are since they were looking for a parser combinator to use, then the additional explanation probably isn't necessary.